### PR TITLE
Allow to create QuadraticModels of other Float types

### PR DESCRIPTION
**Pull request by geoffroyleconte**

### CODE REVIEW

File: src/qpmodel.jl [hunk 1]
```diff
@@ -21,10 +21,10 @@ end
...
 function QuadraticModel(c :: AbstractVector,
                         Hrows :: AbstractVector{<: Integer}, Hcols :: AbstractVector{<: Integer}, Hvals :: AbstractVector;
-                        Arows :: AbstractVector{<: Integer} = Int[], Acols :: AbstractVector{<: Integer} = Int[], Avals :: AbstractVector = Float64[],
-                        lcon :: AbstractVector = Float64[], ucon :: AbstractVector = Float64[],
-                        lvar :: AbstractVector = fill(-Inf, length(c)), uvar :: AbstractVector = fill(Inf, length(c)),
-                        c0 :: Float64=0.0, kwargs...)
+                        Arows :: AbstractVector{<: Integer} = Int[], Acols :: AbstractVector{<: Integer} = Int[], Avals :: AbstractVector = eltype(c)[],
+                        lcon :: AbstractVector = eltype(c)[], ucon :: AbstractVector = eltype(c)[],
+                        lvar :: AbstractVector = fill(eltype(c)(-Inf), length(c)), uvar :: AbstractVector = fill(eltype(c)(Inf), length(c)),
+                        c0 :: T = zero(eltype(c)), kwargs...) where T
```
Comment (abelsiqueira) [inline src/qpmodel.jl:27]: You can define `c :: AbstractVector{T}` and use `T` instead of `eltpye(c)`.

File: src/qpmodel.jl [hunk 2]
```diff
@@ -50,9 +50,10 @@ function QuadraticModel(c :: AbstractVector,
 end
 
 function QuadraticModel(c :: AbstractVector, H :: AbstractMatrix;
-                        A :: AbstractMatrix=zeros(0,length(c)), lcon :: AbstractVector=zeros(0), ucon :: AbstractVector=zeros(0),
-                        lvar :: AbstractVector = fill(-Inf, length(c)), uvar :: AbstractVector = fill(Inf, length(c)),
-                        c0 :: Float64=0.0, kwargs...)
+                        A :: AbstractMatrix=zeros(eltype(c), 0,length(c)),
+                        lcon :: AbstractVector=zeros(eltype(c), 0), ucon :: AbstractVector=zeros(eltype(c), 0),
+                        lvar :: AbstractVector = fill(eltype(c)(-Inf), length(c)), uvar :: AbstractVector = fill(eltype(c)(Inf), length(c)),
+                        c0 :: T = zero(eltype(c)), kwargs...) where T
```
Comment (abelsiqueira) [inline src/qpmodel.jl:56]: Same here.

File: src/qpmodel.jl [hunk 3]
```diff
@@ -64,7 +65,8 @@ function QuadraticModel(c :: AbstractVector, H :: AbstractMatrix;
     nnz(A), findnz(A)...
   else
     I = ((i,j,A[i,j]) for i = 1:ncon, j = 1:nvar)
-    nvar * ncon, getindex.(I, 1)[:], getindex.(I, 2)[:], getindex.(I, 3)[:]
+    nvar * ncon, convert(Array{Int, 1}, getindex.(I, 1)[:]),
```
Comment (abelsiqueira) [inline src/qpmodel.jl:68]: Why is this needed?

**Review by abelsiqueira — CHANGES_REQUESTED** [review]
Thanks, I made a few comments.
I also wonder if we should set all types of `QPData` to `T` (`c :: Vector{T}`, etc.).
The last thing is this has a conflict. Our workflow involves some rebasing to keep the history clean, so probably you forgot to merge `jso/master` into your `master` before branching out - a common situation. I've been recommending GitKraken as a GUI to fix this, but it can be done solely on CLI. Let me know if you don't figure out how.

File: src/qpmodel.jl (hunk 3 above)
Comment (geoffroyleconte) [reply src/qpmodel.jl:68]: Thank you for your comments.
I don't really know why it is needed, but when A is empty, `getindex.(I, 1)[:]` is `Any[]`. The error message is:

`ERROR: LoadError: LoadError: MethodError: no method matching QuadraticModels.QPData(::Float64, ::Array{Float64,1}, ::Array{Int64,1}, 
::Array{Int64,1}, ::Array{Float64,1}, ::Array{Union{Float64, Int64},1}, ::Array{Union{Float64, Int64},1}, ::Array{Union{Float64, Int64},1})
Closest candidates are:
  QuadraticModels.QPData(::T, ::Array{T,1}, ::Array{Int64,1}, ::Array{Int64,1}, ::Array{T,1} where T, ::Array{Int64,1}, ::Array{Int64,1}, ::Array{T,1} where T) where T at C:\Users\Geoffroy Leconte\.julia\dev\QuadraticModels.jl\src\qpmodel.jl:4`

I am sorry for the github issue, does merging jso/master into this branch will solve it? (I work with git CLI)

File: src/qpmodel.jl (hunk 2 above)
Comment (dpo) [reply src/qpmodel.jl:56]: Should we use a parametric type then, e.g., `QuadraticModel{T}`? That would mean that we also need `AbstractNLPModels{T}`.

File: src/qpmodel.jl (hunk 3 above)
Comment (abelsiqueira) [reply src/qpmodel.jl:68]: About the git issue, nope. I created a walkthrough here: https://asciinema.org/a/361816 - The screen was cut, the graph command is `git log --oneline --graph --all --decorate`.

Some guidance for the future:
- Your master is never leading, always following. So after the PR is merged, use `git fetch --all` to update your remotes, `git chekout master` and `git merge jso/master --ff-only` to update your master.
- Go to the master, branch, create a/some commit(s) and create a pull request for that branch
- Repeat as needed, always branching from the master
- If the pull requests are "orthogonal", we can merge them without conflicts, otherwise you're gonna have to rebase (see below)
- If some pull requests conflicted, you're gonna have to rebase your conflicted branch to the new master. Say `issueA` can't be merged and we ask you to rebase. Update the master as above then
  - `git rebase master issueA`
  - Conflicts will arise. Open your editor and fix them.
  - `git add files-that-had-conflict`
  - `git rebase --continue`
  - The larger the commit difference, the larger the conflict.
- When a branch is merged, we use rebase or squash, which means that your branch information is used, but not the actual commit. You can safely delete your branch because the master is now updated.

Comment (abelsiqueira) [reply src/qpmodel.jl:68]: I think it's better to add a `if ncon == 0` special case in line 64:
```
nnzj, Arows, Acols, Avals = if ncon == 0
    0, Int[], Int[], T[]
elseif issparse(A)
    nnz(A), findnz(A)...
```

File: src/qpmodel.jl (hunk 2 above)
Comment (abelsiqueira) [reply src/qpmodel.jl:56]: It is not required for more abstract types to have parametric types, to have parametric types on the concrete types.
And parametric types on NLPs might break mixed-precision methods. Also, we can argue that `f(x) = x^2` is not tied to a type, but QPData is tied to arrays that have `eltype`s, so it is tied to those types. The question is more on whether we want to accept `eltype(A) = Float32` and `eltype(Q) == Float64` at the time or not.

Comment (dpo) [reply src/qpmodel.jl:56]: We don't.

Comment (abelsiqueira) [reply src/qpmodel.jl:56]: Then we can use use `T` for every array. :+1:

File: src/qpmodel.jl (hunk 3 above)
Comment (dpo) [reply src/qpmodel.jl:68]: @abelsiqueira It would be great to add such a guide on git branching to our org's wiki.

Comment (geoffroyleconte) [reply src/qpmodel.jl]: Thank you! I hope it is ok now.

File: src/qpmodel.jl [hunk 4]
```diff
@@ -1,8 +1,8 @@
...
 
-mutable struct QPData
-  c0    :: Float64          # constant term in objective
-  c     :: Vector           # linear term
+mutable struct QPData{T}
+  c0    :: T          # constant term in objective
+  c     :: Vector{T}           # linear term
   Hrows :: Vector{Int}      # quadratic term
   Hcols :: Vector{Int}
   Hvals :: Vector
```
Comment (abelsiqueira) [inline src/qpmodel.jl:8]: Following our discussion on types, `Hvals` and `Avals` should be `Vector{T}`.

**Comment by abelsiqueira** [comment]
@geoffroyleconte, the commits are fixed, thanks!

File: src/qpmodel.jl [hunk 5]
```diff
@@ -49,18 +49,21 @@ function QuadraticModel(c :: AbstractVector,
                  QPData(c0, c, Hrows, Hcols, Hvals, Arows, Acols, Avals))
 end
 
-function QuadraticModel(c :: AbstractVector, H :: AbstractMatrix;
-                        A :: AbstractMatrix=zeros(0,length(c)), lcon :: AbstractVector=zeros(0), ucon :: AbstractVector=zeros(0),
-                        lvar :: AbstractVector = fill(-Inf, length(c)), uvar :: AbstractVector = fill(Inf, length(c)),
-                        c0 :: Float64=0.0, kwargs...)
+function QuadraticModel(c :: AbstractVector{T}, H :: AbstractMatrix;
+                        A :: AbstractMatrix=zeros(T, 0,length(c)),
```
Comment (dpo) [inline src/qpmodel.jl:53]: space missing here.

**Review by dpo — APPROVED** [review]
Just a final minor comment.

File: src/qpmodel.jl [hunk 6]
```diff
@@ -50,7 +50,7 @@ function QuadraticModel(c :: AbstractVector{T},
 end
 
 function QuadraticModel(c :: AbstractVector{T}, H :: AbstractMatrix;
-                        A :: AbstractMatrix=zeros(T, 0,length(c)),
+                        A :: AbstractMatrix=zeros(T, 0, length(c)),
                         lcon :: AbstractVector=zeros(T, 0), ucon :: AbstractVector=zeros(T, 0),
```
Comment (dpo) [inline src/qpmodel.jl:54]: Sorry for being annoying. There are a few other spaces missing around `=` (to be consistent everywhere).

**Review by abelsiqueira — APPROVED** [review]

**Comment by dpo** [comment]
Thank you @geoffroyleconte !
